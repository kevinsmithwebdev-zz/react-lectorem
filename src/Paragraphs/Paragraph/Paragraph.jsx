import React from 'react';
import Lect from './Lect/Lect';

export const TIME_CODE_FUDGE = 0.1;

export const getShouldHighlight = (lect, readTime) =>
  lect.start - TIME_CODE_FUDGE <= readTime &&
  readTime <= lect.end + TIME_CODE_FUDGE;

export const renderLect = (lect, idx, readTime, showModal) => {
  const shouldHighlight = getShouldHighlight(lect, readTime);
  const key = idx;
  return (
    <Lect
      lect={lect}
      shouldHighlight={shouldHighlight}
      key={key}
      readTime={readTime}
      showModal={showModal}
    />
  );
};

const Paragraph = ({
  paragraph: { header, lects } = {},
  idx,
  readTime,
  showModal,
}) => (
  <div className="paragraph" key={idx}>
    <div className="Lectorem-paragraph-header">
      {!!header?.imagePath && (
        <div>
          <img src={header?.imagePath} alt="Logo" width="150px" />
        </div>
      )}
      {header?.lects?.map(
        /* istanbul ignore next */ (lect, idx2) =>
          renderLect(lect, idx2, readTime, showModal)
      )}
    </div>
    <div>
      {lects?.map(
        /* istanbul ignore next */ (lect, idx2) =>
          renderLect(lect, idx2, readTime, showModal)
      )}
    </div>
  </div>
);

export default Paragraph;
