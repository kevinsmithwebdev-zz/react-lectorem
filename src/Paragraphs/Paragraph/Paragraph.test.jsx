import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Paragraph, {
  getShouldHighlight,
  renderLect,
  TIME_CODE_FUDGE,
} from './Paragraph';
import storyData from '../../../fixture/storyData.json';

const mockParagraph = storyData.paragraphs[0];

describe('Paragraph', () => {
  describe('the main component render', () => {
    const renderer = new ShallowRenderer();
    describe('empty', () => {
      it('should render without crashing', () => {
        renderer.render(<Paragraph />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });

    describe('with data', () => {
      const props = {
        paragraph: mockParagraph,
        idx: 123,
        readTime: 12.34,
        showModal: 'showModal123',
      };

      it('should render and match snapshot', () => {
        renderer.render(<Paragraph {...props} />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });
  });

  describe('getShouldHighlight', () => {
    const lect = {
      start: 1000,
      end: 2000,
    };

    describe('when readTime is well before start', () => {
      const readTime = lect.start - TIME_CODE_FUDGE * 2;
      it('should return false', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(false);
      });
    });

    describe('when readTime is before start but within fudge', () => {
      const readTime = lect.start - TIME_CODE_FUDGE / 2;
      it('should return true', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(true);
      });
    });

    describe('when readTime is well within start and stop', () => {
      const readTime = lect.start + (lect.end - lect.start) / 2;
      it('should return true', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(true);
      });
    });

    describe('when readTime is after start but within fudge', () => {
      const readTime = lect.end + TIME_CODE_FUDGE / 2;
      it('should return true', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(true);
      });
    });

    describe('when readTime is well after end', () => {
      const readTime = lect.end + TIME_CODE_FUDGE * 2;
      it('should return false', () => {
        expect(getShouldHighlight(lect, readTime)).toBe(false);
      });
    });
  });

  describe('renderLect', () => {
    const lect = mockParagraph.lects[0];
    const idx = 12;
    const readTime = 123;
    const showModal = 'showModal123';
    it('should render and match snapshot', () => {
      expect(renderLect(lect, idx, readTime, showModal)).toMatchSnapshot();
    });
  });
});
