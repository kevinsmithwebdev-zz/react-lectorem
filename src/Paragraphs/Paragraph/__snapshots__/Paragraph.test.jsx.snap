// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph renderLect should render and match snapshot 1`] = `
<Lect
  lect={
    Object {
      "end": 3.22,
      "explanation": "This is how we say hello.",
      "start": 2.64,
      "text": "Hola.",
    }
  }
  readTime={123}
  shouldHighlight={false}
  showModal="showModal123"
/>
`;

exports[`Paragraph the main component render empty should render without crashing 1`] = `
<div
  className="paragraph"
>
  <div
    className="Lectorem-paragraph-header"
  />
  <div />
</div>
`;

exports[`Paragraph the main component render with data should render and match snapshot 1`] = `
<div
  className="paragraph"
>
  <div
    className="Lectorem-paragraph-header"
  >
    <div>
      <img
        alt="Logo"
        src="./burro.png"
        width="150px"
      />
    </div>
    <Lect
      lect={
        Object {
          "end": 1.18,
          "start": 0.36,
          "text": "El burro",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 2.58,
          "explanation": "The Spanish word \\"bailarín\\" can be both and adjective (\\"dancing\\") and a noun (\\"dancer\\").",
          "start": 1.18,
          "text": "bailarín.",
          "translation": "dancer",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
  </div>
  <div>
    <Lect
      lect={
        Object {
          "end": 3.22,
          "explanation": "This is how we say hello.",
          "start": 2.64,
          "text": "Hola.",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 4.12,
          "explanation": "",
          "start": 3.6,
          "text": "Soy",
          "translation": "I am",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 4.98,
          "explanation": "",
          "start": 4.12,
          "text": "un burro.",
          "translation": "a donkey.",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 6.36,
          "explanation": "",
          "start": 5.42,
          "text": "Me llamo",
          "translation": "My name is",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 7.04,
          "explanation": "",
          "start": 6.36,
          "text": "Beto.",
          "translation": "Beto.",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 8.36,
          "explanation": "",
          "start": 7.7,
          "text": "Tengo",
          "translation": "I have",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 8.86,
          "explanation": "",
          "start": 8.4,
          "text": "botas.",
          "translation": "boots.",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 9.92,
          "explanation": "",
          "start": 9.46,
          "text": "Tengo",
          "translation": "I have",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 10.52,
          "explanation": "",
          "start": 9.92,
          "text": "botas",
          "translation": "boots",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
    <Lect
      lect={
        Object {
          "end": 11.6,
          "explanation": "",
          "start": 10.68,
          "text": "para bailar.",
          "translation": "for dancing.",
        }
      }
      readTime={12.34}
      shouldHighlight={false}
      showModal="showModal123"
    />
  </div>
</div>
`;
